GitHub actions

Bartleby2718 opened the pull request:
# Status
Done

# What I did
- Finished GitHub Actions (btw I'm kinda hesitant to use the acronym GA because it's often used for Google Analytics, which we also use.)

## Issues addressed
- Supersedes PR #331 Env vars 
- Closes #147 "Switch from Travis to GitHub Actions"
- Closes #329 "Use a different settings.py for each environment"
- Closes #347 "ESLint doesn't work"
- Invalidates / Closes #121 "Move Docker image to a registry to speed up builds on Travis CI" 

## Screenshots
1. When there is an update to a PR
![image](https://user-images.githubusercontent.com/7676354/104119182-290c1680-5371-11eb-82c0-a0fe7ce4a05e.png)
![image](https://user-images.githubusercontent.com/7676354/104119371-2cec6880-5372-11eb-9369-d93319de6601.png)

2. When there is a push to the selected branches
![image](https://user-images.githubusercontent.com/7676354/104119194-375a3280-5371-11eb-8093-ea31c89cdc98.png)
![image](https://user-images.githubusercontent.com/7676354/104119360-1fcf7980-5372-11eb-87a2-6e85f7d197b8.png)

j-alicia-long commented:
Haha could always use GHA? Ghana Airways lol

This is great though - let us know if you need input/reviews/testing on anything!

Bartleby2718 commented:
To-Do (after merge):
- Update environment variables on Heroku once this is merged to `master`
- Disable TravisCI once this is merged to `master`
- Things to update on Google Drive: `Deployment`, `theCourseForum 2.0 Secrets`, `environment variables` folder

alex-shen1 commented on tcf_core/settings/testing.py:
```diff
@@ -0,0 +1,18 @@
+# pylint: disable=import-error,unused-wildcard-import,wildcard-import
+"""Django settings module for testing environment such as Heroku"""
```
I feel like we've had this discussion somewhere before (don't remember where, so I'll say it again), but I think the name of this file ought to be something like `dev` or maybe `dev-build` for clarity at the expense of conciseness; `testing.py` sounds like a scratch file used for experimenting with the settings themselves.

Bartleby2718 replied:
Bump in case you haven't checked my last comment: https://github.com/thecourseforum/theCourseForum2/pull/331#discussion_r544808854.
I can change it if you still think we should go with `dev` after reading this.

alex-shen1 replied:
Ah, knew it was somewhere. I see your point - I still feel like since I'm pretty sure the term "dev build" is pretty standard, we should go with that, but maybe something like `test-build` would work too? Just averse to the word `testing` specifically. Probably should get input from others - @j-alicia-long @jackliu612  ?

Bartleby2718 replied:
I've thought more about this and found more reasons to go with `dev`:
1. Our default branch is `dev`.
2. The URLs are `dev.thecourseforum.com` and `thecourseforum-dev.herokuapp.com`.
3. The order of the files matches the flow of the source code integration:
![image](https://user-images.githubusercontent.com/7676354/106097636-abccf980-6105-11eb-8e33-fdd77c1a8281.png)

Hence, fixed in 489f15807ce4cd3dca935c52a8f000a06d08bf3f.

alex-shen1 commented on tcf_website/templates/base/base.html:
```diff
@@ -69,6 +69,7 @@
     return "";
   }
   // Update copyright year in the sidebar
+  // Adding a comment to test if deployment succeeded
```
Should remove this once we merge it in. Also, I can't find this comment on the site source, so maybe it didn't work?

Bartleby2718 replied:
Since this branch isn't merged to `dev` (or `master`), you won't see it anywhere else besides the local server launched in the `github-actions` branch. I know this is not a good way to check if deployment worked, but I thought a single line of comment won't hurt.

alex-shen1 replied:
If the concern is checking whether or not deploy worked, I think for `dev` we have Heroku logs, and the next time `master` gets updated there'd be a lot of other features added that we can use as indicators? (also GAE logs, but I don't really know where they are lol)

alex-shen1 replied:
If you think it's easier to just leave in the comment for now, just don't forget to remove it once we verify both deployments work!

Bartleby2718 replied:
Now that I think about it, this was definitely a dumb mistake that I made when testing GitHub Actions by manually pushing to `master`. Fixed in bfb73944e1775b45e4bdead29d4b82b913f18673.

alex-shen1 reviewed (changes requested):
Good work! Seems like this'll help with maintainability in the future. Have a few comments/questions:

Bartleby2718 commented on .github/workflows/continuous-integration.yml:
```diff
@@ -2,67 +2,111 @@
+          python-version: ${{ env.PYTHON_TARGET }}
 
-      - name: Install Python packages
+      - name: Install Python packages, excluding the unnecessary ones
```
> By default, a workflow with multiple jobs will run those jobs in parallel. 

https://docs.github.com/en/actions/learn-github-actions/introduction-to-github-actions#jobs
The `django` job isn't _reinstalling_ the packages--it is installing the packages _in a different container_. I wanted to minimize the total build time, although It won't make much of a difference.

I could have combined the two (or three) jobs into one, but in that case, we might need to wait for another run if the build failed in the first step.

alex-shen1 replied:
Right, I used "reinstall" as a shorthand; I should've been more clear. The page you linked suggests that all these installations are run in parallel anyways (and since we're a public repo, I think we don't need to worry about paying for compute power?), so it seems my previous concern doesn't apply.

I believe Travis builds would run all tests even if a step failed; is this not the case for Actions?

Bartleby2718 replied:
You can use `continue-on-error` to keep going, but you would have to manually check the result to see where you failed, whereas right now you can see on Discord which job failed.

More importantly, I think the three checks are semantically different and independent enough to be run as separate jobs.

Bartleby2718 replied:
@alex-shen1 I combined `pylint` and `django` in 6e43fb2f69de7c33ff59410cc50d33324bf699e1. Does this look better? (It's definitely shorter.)

alex-shen1 replied:
Looks good to me!

alex-shen1 replied:
On second thought, the name `django` doesn't exactly convey that you're running Pylint in the job. I'm guessing that you merged those because they need Python stuff installed, so maybe the name `python` would work, but that isn't fantastic either. The other alternative I can think of is moving Pylint to `eslint` and renaming the job something like `linting`, but that sorta defeats the point of merging jobs.

Bartleby2718 replied:
LOL I'll just split them back then.

Bartleby2718 replied:
Reverted in 01115bab815a28416d7ddd10cca0922f0d072a3d

Bartleby2718 commented on tcf_core/settings/ci.py:
```diff
@@ -0,0 +1,13 @@
+# The following if statement is needed to prevent overwriting global variables when
+# this settings file is interpreted
+if os.environ.get('DJANGO_SETTINGS_MODULE') == 'tcf_core.settings.ci':
+    pass
```
Uhh, it just so happens that `settings/ci.py` coincides with `settings/base.py` because environment variables were enough to handle all environment-specific differences, but the `if` statement can be used to make any changes to Django settings without using environment variables. (If this `if` statement were absent, future maintainers will encounter the same bug that I experienced.) Also, it makes the settings modules consistent.

Bartleby2718 commented on scripts/notify-discord.sh:
```diff
@@ -0,0 +1,50 @@
+function get_content {
+    case $GITHUB_EVENT_NAME in 
+      pull_request)
+        echo "**[PR #$PR_NUMBER: $PR_TITLE](https://github.com/thecourseforum/theCourseForum2/pull/$PR_NUMBER)** ([\`$GITHUB_HEAD_REF\`](https://github.com/thecourseforum/theCourseForum2/tree/$GITHUB_HEAD_REF)->[\`$GITHUB_BASE_REF\`](https://github.com/thecourseforum/theCourseForum2/tree/$GITHUB_BASE_REF))" ;;
```
Fixed in de6df3cce531294a84b03d3fed6e5d8754a247f2

Bartleby2718 commented on docs/FOR-DEVS.md:
```diff
@@ -171,4 +161,4 @@ python3 manage.py migrate
 
 ### Deploying to prod
-Production deployments (i.e. the version of the site users see on thecourseforum.com) should be handled automatically by our CI (continuous integration system). As of the writing of this documentation, Travis is set up to autodeploy every time a build on the `master` branch passes and we're working on moving towards GitHub Actions. However, in the event that something fails, the manual deploy pro ...
+Production deployments (i.e. the version of the site users see on thecourseforum.com) should be handled automatically by our CI (continuous integration system). As of the writing of this documentation, GitHub Actions is set up to autodeploy every time a build on the `master` branch passes and we're working on moving towards GitHub Actions. However, in the event that something fails, the manual de ...
```
Fixed in ee68b13c518b1d00550b5e3c2fad23b97db14867

Bartleby2718 commented on .github/workflows/continuous-integration.yml:
```diff
@@ -2,67 +2,111 @@
+        if: ${{ success() }}
+        id: coverage
+        run: |
+          echo "::set-output name=percentage::$(coverage report | grep -o '[0-9]\+%' | tail -1)"
```
`{0}` seems like an instance of [brace expansion](https://www.gnu.org/software/bash/manual/html_node/Brace-Expansion.html). Just for your information, `grep -o '[0-9]+\%'` finds 1 or more decimal digits followed by a percent sign, and `tail -1` returns only the last occurrence.

alex-shen1 replied:
Right, but isn't the point of `echo` to print out the number? It doesn't appear to be part of the workflow log. In any case, the number seems to be getting passed to the Discord output so there's definitely something there (71% coverage seems a bit high though considering the low number of tests we have).

Bartleby2718 replied:
I'm not sure if we should be looking for the output of `echo` in the log. The command is there because [that's how output is set, according to the GitHub Actions documentation](https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-an-output-parameter).

Bartleby2718 commented on .github/workflows/continuous-integration.yml:
```diff
@@ -2,67 +2,111 @@
+      - name: Install Python packages, excluding the unnecessary ones
+        run: |
+          python -m pip install --upgrade pip
+          sed -i '/\(tqdm\|gunicorn\|django\-heroku\|uwsgi\)/d' requirements.txt
```
Hmm, I do agree that this is ugly code (just like other code for optimization), but I can't think of a good alternative (besides taking it out). Personally, I don't think it'll change that often, so I'd like to keep the three `sed` commands and reduce build time.

Bartleby2718 replied:
By the way, there's [this approach](https://stackoverflow.com/a/20720019), although [Heroku doesn't recommend it](https://devcenter.heroku.com/articles/python-pip#cascading-requirements-files).

alex-shen1 replied:
Yeah, I think it's probably not worth the hassle of introducing new files just for the sake of a marginal amount of code cleanliness. My guess is that even if we have new packages we *can* ignore, it doesn't really hurt if the hardcoded settings don't get updated until later.

Bartleby2718 commented on tcf_core/settings/base.py:
```diff
@@ -174,53 +159,22 @@
-        'PORT': env.str('DB_PORT'),
-        'OPTIONS': {'sslmode': 'require'},
-    }
+# Read-write access to Elastic
```
Didn't know that, so I fixed it in 59c80668af703ff484fe54a99b94a9bfb2db9380 (and updated Google Drive accordingly).

Bartleby2718 commented on scripts/notify-discord.sh:
```diff
@@ -0,0 +1,50 @@
+EOF
+)
+
+curl $DISCORD_WEBHOOK \
```
Fixed in 9b207ead54a1969712c0010fdfe4062f36fd5b51, following @j-alicia-long 's comment (on the Discord server)

alex-shen1 commented on .github/workflows/master-to-app-engine.yml:
```diff
@@ -69,7 +69,7 @@ jobs:
 
       - name: Substitute environment variables
-        run: envsubst < .config/app.yaml.template > app.yaml
+        run: envsubst < .config/app.template.yaml > app.yaml
```
Is this distinction important to make the deploy work (i.e. do you think the approach before wouldn't work)? We have a `.env.template` as opposed to `.template.env`, so for consistency I feel like the `.template` being at the end sounds better. Kinda pedantic though, so if there's a reason then this point is moot.

Bartleby2718 replied:
You're not being pedantic at all! All comments are valuable.

I had not thought about `.env` files, but now that you've mentioned it, we can rename them to ensure proper extensions are used. (After all, this PR already changes a lot about environment variables anyways.) I would prefer using prefixes or infixes, since suffixes should be used for extensions (which, in turn, are used by IDEs for color coding and auto-formatting. Do you mind if I change as such?

alex-shen1 replied:
I'm not sure.... from my brief investigation, it seems as if `.env.template` seems to be the more common approach. In fact, if you Google exactly `".template.env"`, the [first page that comes up](https://dump-env.readthedocs.io/en/latest/) is about using `.env.template` files. However, adding to the confusion, it does mention `.template.env` files too, but those _appear_ to be used in a different context. I'm not sure that syntax highlighting is _terribly_ important for `.env` files considering they're just a series of `{NAME}={VALUE}` lines either.

Regardless on what we end up deciding, we should definitely strive for consistency.

Bartleby2718 replied:
What I want is syntax highlighting for `app.yaml` because there is a huge difference:
![colorcodedyaml](https://user-images.githubusercontent.com/7676354/106096555-d027d680-6103-11eb-9d0a-7c6983147eef.png)
![plaintextyaml](https://user-images.githubusercontent.com/7676354/106096556-d027d680-6103-11eb-967c-b30020c2679f.png)

How about we make a compromise and use suffixes for plaintext files (which don't need formatting) and infixes for others? 🤔

Bartleby2718 replied:
Reverted in 0c42f0367937333ecca53b3baf4561e23edc29c4 for now.

alex-shen1 commented on .github/workflows/continuous-integration.yml:
```diff
@@ -81,18 +66,33 @@ jobs:
           python -m pip install --upgrade pip
-          sed -i '/\(pandas\|tqdm\|lint\|autopep8\|gunicorn\|django\-heroku\|uwsgi\)/d' requirements.txt
+          sed -i '/\(pandas\|tqdm\|gunicorn\|django\-heroku\|uwsgi\)/d' requirements.txt
+          echo coverage >> requirements.txt
```
Would it be good to add `coverage` into `requirements.txt` directly, so coverage can be run as a part of `precommit`? (perhaps outside the scope of this PR)

Bartleby2718 replied:
That's actually a great idea! I'll work on that as soon as this PR is merged.

Bartleby2718 replied:
Created an issue for this: #361.

Bartleby2718 commented:
@alex-shen1 Do you think [this](https://docs.github.com/en/github/finding-security-vulnerabilities-and-errors-in-your-code/configuring-code-scanning) will help?

alex-shen1 commented:
@Bartleby2718 Sorry for taking a while to respond, it looks cool! The one thing that's concerning is just how many checks it adds — I never understood if there's a difference between `pull_request` and `push` checks, but do we really need 4 checks just for code scanning? Not terribly important, but I'm not a huge fan of how it clutters the checks list (especially when one of the CI ones is sandwiched inbetween CodeQL checks..?). 

Also, the "code scanning results" check doesn't seem to be entirely working? It just says "2 analyses not found".

alex-shen1 commented on .github/workflows/codeql-analysis.yml:
```diff
@@ -0,0 +1,32 @@
+  pull_request:
+    branches: [ dev ]
+  schedule:
+    - cron: '17 9 * * 6'
```
If this is correct: 
![image](https://user-images.githubusercontent.com/55100084/106558429-b3b7df80-64f1-11eb-8d8c-30b1685b9bd4.png)

You want this to run at 9:17 every Sunday..,? Is the idea here that vulnerabilities might update over time, so we need to recheck regularly? Assuming I've understood correctly, I feel like pushes/PRs from normal dev activity would be regular enough to catch updates when those get checked.

Bartleby2718 replied:
I believe I just copied and pasted most of the things. I just checked with another repository, and it seems like the time is randomly chosen. I'll get rid of `cron` if we do decide to add this.

alex-shen1 reviewed (commented):
Looks pretty solid so far - is this still WIP, or are you done? Because I noticed you requested a review earlier, but then started working on code scanning.

In any case, looks like we only have [one problem](https://github.com/thecourseforum/theCourseForum2/security/code-scanning?query=ref%3Arefs%2Fpull%2F344%2Fmerge+tool%3ACodeQL) caught by the scanner? Still learning how all this is working.

Bartleby2718 commented on .github/workflows/codeql-analysis.yml:
```diff
@@ -0,0 +1,32 @@
+
+on:
+  push:
+    branches: [ dev, master, github-actions ]
```
Yeah, I noticed that no checks were run initially when I didn't have `github-actions` in `branches`, but now I see that there are duplicate checks. Still, I'm not sure one check for Python and another for JavaScript is necessary.

Bartleby2718 commented:
> Also, the "code scanning results" check doesn't seem to be entirely working? It just says "2 analyses not found".

I think it's because this it's only implemented in the head branch (`github-actions`) and not in the base branch (`dev`)
![image](https://user-images.githubusercontent.com/7676354/106584734-8cbed500-6514-11eb-85e1-09365b902b58.png)

> Looks pretty solid so far - is this still WIP, or are you done? Because I noticed you requested a review earlier, but then started working on code scanning.

I requested a review as soon as I was done with GitHub Actions, but I came upon code scanning afterwards, and it requires GitHub Actions, so I just added it to this PR. Sorry for the confusion.
> In any case, looks like we only have one problem caught by the scanner?

Yes, I guess that means our code was pretty secure.

@alex-shen1 I don't have strong opinions on this, so I can just take it out if you don't find this necessary. I just added this because it was worth looking into (although perhaps not incorporating). To be honest, I would rather have the original PR get merged ASAP if significant changes need to be made on code scanning because this PR has been open for too long.

alex-shen1 commented:
> To be honest, I would rather have the original PR get merged ASAP if significant changes need to be made on code scanning because this PR has been open for too long.

Yeah, I was gonna say the same thing but didn't want to invalidate any of the work you've done. You could cherry-pick all all the CodeQL commits off into a separate branch and we can just get this merged?

alex-shen1 reviewed (approved):
Finally, we're all good to go!
